### ISSUE TITLE
numberOfRecordsReturned does not reflect actual number of records returned in all cases

### ISSUE THREAD

**Issue by rjaduthie** [geopython/pycsw#556]
# Description
From Table 66 (p159) of 07-006r1 OpenGIS Catalogue Services Specification v2.0.2:

>numberOfRecordsReturned :
>
>Number of records actually returned to client. This may not be the entire result set since some servers may limit the number of records returned to limit the size of the response package transmitted to the client. Subsequent queries may be executed to see more of the result set. The nextRecord attribute will indicate to the client where to begin the next query.

However, when 'startposition' and 'maxrecords' parameters  total more than the number of 'matched' records in a catalogue, the 'returned' variable is set to 'maxrecords' and not ('matched' - 'startposition' + 1).

From pycsw/ogc/csw/csw2.py [MASTER] (lines 878-890), with added comment:

```python
if int(matched) == 0:
    returned = nextrecord = '0' # OK, makes sense
else:
    if int(matched) < int(self.parent.kvp['maxrecords']):
        returned = matched # OK, makes sense
        nextrecord = '0'
    else:
        returned = str(self.parent.kvp['maxrecords']) # Here is the problem.
        if int(self.parent.kvp['startposition']) + int(self.parent.kvp['maxrecords']) >= int(matched):
            nextrecord = '0'
        else:
            nextrecord = str(int(self.parent.kvp['startposition']) + \
            int(self.parent.kvp['maxrecords']))
```

There is similar code in pycsw/ogc/csw/csw/csw3.py [MASTER].

The 'returned' value is then used to specify the 'numberOfRecordsReturned' attribute in the response.

# Environment
Confirmed in on 1.10.1 (see test below).
The code which controls this behaviour is in MASTER in the respository for both CSW2 and CSW3.

# Steps to Reproduce
e.g. Try this URL in a browser:
<https://api.bas.ac.uk/data/metadata/csw/v1?service=CSW&version=2.0.2&request=GetRecords&elementsetname=full&typenames=csw:Record&startposition=345&resulttype=results>

You'll notice that the numberOfRecordsReturned attribute in the csw:SearchResults is 100, although there are only 5 child csw:Record elements.

# Additional Information

...

**Comment by ricardogsilva** [comment]
@rjaduthie thanks for this

I'll do a review over the weekend and come back with some feedback

**Comment by rjaduthie** [comment]
N.B. This issue already has an associated pull request, addressing the issue.  @ricardogsilva I think you're aware, but there's no harm in making sure.

**Comment by tomkralidis** [comment]
Thanks @rjaduthie / @ricardogsilva.  Relevant PR is in #557

**Comment by tomkralidis** [comment]
#557 merged into master and 2.2 branch.  Thanks @rogie-bas!

**Comment by mikejmets** [comment]
If #557 is merged, shouldn't this issue be closed?
